---
title: "R Notebook COOOL Population Structure"
author: "Thomas Chevrier"
output:
  pdf_document: default
  html_notebook: default
---

The first step is to install and load all the libraries we need to run the entire script.  

```{r, include=FALSE, echo=FALSE, warning=FALSE}
library(Rcpp) 
library(remotes)
library(atease) 
library(mvbutils)
library(vecless)
library(SNPRelate)
library(gbasics) # For any questions or enquiries about package gbasics contact the package maintainer Shane Baylis at Shane.Baylis@csiro.au
library(kinference) # For any questions or enquiries about package kinference contact the package maintainer Shane Baylis at Shane.Baylis@csiro.au
library(tidyr)
library(stringr)
library(tidyverse)
library(dplyr)
library(janitor)
library(dartR)
library(adegenet)
library(ape)
library(hierfstat)
library(reshape2)
library(pcadapt)
library(qvalue)
library(poppr)
library(graph4lg)
library(ggnewscale)
library(data.table)
```

The next step is to import the sequencing data, here we are working with the report sent by DArT. However before importing the report a metadata file must be done including at least two column *pop* and *id* such as below. Both files will be load together in R.

```{r, include=FALSE, echo=FALSE, warning=FALSE}
metadata <- read.csv(file = "my_ind_metric_filage.csv")
```


```{r}
head(metadata)
```

Once the metadata file is made, the importation can be done using the *gl.read.dart* function. The report could have different format (2 row or 1 row). For more information you can check this tutorial made by Gruber et al. http://georges.biomatix.org/storage/app/media/uploaded-files/tutorial3adartrdatastructuresandinput22-dec-21-2.pdf


```{r, include=TRUE, echo=FALSE, warning=FALSE,}
gl_dapc <- gl.read.dart(filename="Report_DXi22-7046_SNP_mapping_tot.csv", ind.metafile = "my_ind_metric_filage.csv")
```


```{r}
head(gl_dapc)
```

The first value correspond to the number of individual and the second on to the SNPs numbers.
```{r}
dim(gl_dapc)
```


# Filtering the data

Once the data are load in R, we have to filter the SNPs and the individuals which are not follow the threshold we will fix.

The first filter we apply to the data is based on the sequencing depth, we will only keep the SNPs with a sequencing depth between 20 and 145.

If you want to visualise how the sequencing depth is represent in your data you can use the function *gl.report.rdepth()*

```{r, echo=FALSE, fig.height = 5, fig.width = 7}
# gl.report.rdepth(gl_dapc)

gl_dapc <- gl.filter.rdepth(gl_dapc, lower = 20, upper = 145)
```


We can check if the filter using rdepth worked by checking the dimension of our new object 

```{r}
dim(gl_dapc)
```

We saw that the number of individual is unchanged whereas the SNPs number is lower. (If you want you can add dim(gl_dapc) after each filter step to check what you did, the line are added but in comment.)

Then we will only keep on SNPs per sequencing tag to avoid the linkage desequilibrium, for that we run the following code.

```{r}
gl_dapc <- gl.filter.secondaries(gl_dapc)
```


<! -- dim(gl_dapc) --> 

The next step is to filter our data based on the proportion of technical replicate assay pairs for which the marker score is consistent which is called Reproducibility by DArT.

As for rdepth you can use the function *gl.report.reproducibility()* to represent the reproducibility values through the data

```{r, echo=FALSE, fig.height = 5, fig.width = 7}
gl_dapc <- gl.filter.reproducibility(gl_dapc,0.95)
```

<! -- dim(gl_dapc) --> 

Then the filter will focus on the proportion of locus and samples for which the genotype call is either "1" or "0", rather than "‚Äê"
The report can be obtain using gl.report.callrate(,method = "loc") for locus and gl.report.callrate(,method = "ind") for sample.

```{r}
gl_dapc <- gl.filter.callrate(gl_dapc, method = "loc",0.99)
```

<! -- dim(gl_dapc) --> 

```{r}
gl_dapc <- gl.filter.callrate(gl_dapc,method = "ind",0.95)
```

<! -- dim(gl_dapc) --> 

The final filter step using dartR package is based on the Minor allele frequency (maf), as for the other step the report is available using gl.report.maf()

```{r}
gl_dapc <- gl.filter.maf(gl_dapc,0.05) # If your data allow it, keep 0.05 as a threshold is a "reference" data for this type of study
```

```{r}
dim(gl_dapc)
```

```{r}
loci_name_K1_gl_dapc <- gl_dapc@loc.names # Create a vector of Loci name
my_sample_K1_gl_dapc <- gl_dapc@ind.names # Create a vector of individuals name
info_loci_K1_gl_dapc <- gl_dapc@other$loc.metrics # Create a vector of metadata

```

```{r, include=FALSE,echo=FALSE, warning=FALSE}
Ind_info_POP_K1_gl_dapc <- cbind(my_sample_K1_gl_dapc, rep("POP1", 2028))
Ind_info_POP_K1_gl_dapc <- as.data.frame(Ind_info_POP_K1_gl_dapc)
```

For all the Kinference part if you have any question or you want to access to the Kinference package thanks to contact the developper Shane Baylis (Shane.Baylis@csiro.au) 

```{r, include=FALSE,echo=FALSE, warning=FALSE}
my_snpgeno_K1_gl_dapc <- snpgeno.snpgds(x = as.matrix(gl_dapc), Locus = loci_name_K1_gl_dapc, Our_sample = my_sample_K1_gl_dapc, locinfo = info_loci_K1_gl_dapc, info = Ind_info_POP_K1_gl_dapc)

```

```{r,echo=FALSE}
my_snpgeno_K1_gl_dapc[1:10,1:10]

```

```{r, echo=FALSE, fig.height = 7 , fig.width = 10, include=T}
check6and4(my_snpgeno_K1_gl_dapc, thresh_pchisq_6and4=c(0.001, 0.0001))
```

```{r, echo=FALSE, fig.height = 7 , fig.width = 10, include=T}
my_snpgeno_K1_gl_dapc_6_4s <- check6and4(my_snpgeno_K1_gl_dapc, thresh_pchisq_6and4 = c(0.001, 0.00001))
my_snpgeno_K1_gl_dapc_goodLoci <- my_snpgeno_K1_gl_dapc[, my_snpgeno_K1_gl_dapc_6_4s$pval4 > 0.025]
dim(my_snpgeno_K1_gl_dapc_goodLoci)
```

```{r, echo=FALSE, fig.height = 7 , fig.width = 10, include=T}
check6and4(my_snpgeno_K1_gl_dapc_goodLoci, thresh_pchisq_6and4=c(0.001, 0.0001))
```

```{r, include=FALSE,echo=FALSE, warning=FALSE}
my_snpgeno_K1_gl_dapc <- hsp_power(my_snpgeno_K1_gl_dapc_goodLoci, k = 0.5)
my_snpgeno_K1_gl_dapc <- prepare_PLOD_SPA(my_snpgeno_K1_gl_dapc)

```

```{r}
dups <- find_duplicates(my_snpgeno_K1_gl_dapc, max_diff_loci=200, showPlot = TRUE)
```

```{r, echo=FALSE, fig.height = 5, fig.width = 7}
hist(dups$ndiff, xlab = "n different genos", main = "", xlim = c(0,200), nc = 50)
```

```{r, include=FALSE,echo=FALSE, warning=FALSE}
prs <- with( my_snpgeno_K1_gl_dapc$info, cbind(Our_sample[ dups$i], Our_sample[ dups$j], dups,
                             stringsAsFactors=FALSE))

tech.reps <- prs[,1] == prs[,2]

```

```{r, echo=FALSE}
prs
```

```{r, include=FALSE, echo=FALSE, warning=FALSE}
oneofeach <- drop_dups_pairwise_equiv( dups[,2:3])
my_snpgeno_K1_gl_dapc_b <- my_snpgeno_K1_gl_dapc[ -c( oneofeach),]

```

```{r, echo=FALSE, fig.height = 5, fig.width = 7}
Lfish <- ilglk_geno(my_snpgeno_K1_gl_dapc_b, list(nclass=100))
Lthreshs <- c(-1575)
abline(v=Lthreshs, col="red")
```

```{r, include=FALSE, echo=FALSE, warning=FALSE}
keep <- (Lfish > min(Lthreshs))
my_snpgeno_K1_gl_dapc_c <- my_snpgeno_K1_gl_dapc_b[ keep, ]

```

```{r, echo=FALSE, warning=FALSE}
dim(my_snpgeno_K1_gl_dapc_b) - dim(my_snpgeno_K1_gl_dapc_c)

```

```{r, include=FALSE, echo=FALSE, warning=FALSE}
 my_snpgeno_K1_gl_dapc_d <- hsp_power( my_snpgeno_K1_gl_dapc_c, k=0.5)
 my_snpgeno_K1_gl_dapc_d <- prepare_PLOD_SPA(my_snpgeno_K1_gl_dapc_d)

```

```{r, echo=FALSE, fig.height = 5, fig.width = 7}
hetz_rich <- hetzminoo_fancy((my_snpgeno_K1_gl_dapc_d), "rich")
HthreshsRich <- c(0.22, 0.29) ## set the 'extremes'
abline( v=HthreshsRich, col="red")
```


```{r, include=FALSE, echo=FALSE, warning=FALSE}
my_snpgeno_K1_gl_dapc_e <- my_snpgeno_K1_gl_dapc_d[ (hetz_rich > min(HthreshsRich)) &
                  (hetz_rich < max(HthreshsRich)), ]
```


```{r}
dim(my_snpgeno_K1_gl_dapc_d) - dim(my_snpgeno_K1_gl_dapc_e) 
```

```{r, include=FALSE,echo=FALSE, warning=FALSE}
my_snpgeno_K1_gl_dapc_e2 <- re_est_ALF(my_snpgeno_K1_gl_dapc_e)
```


```{r, echo=FALSE, fig.height = 5, fig.width = 7}
hetz_rich <- hetzminoo_fancy(my_snpgeno_K1_gl_dapc_e2, "rich")
HthreshsRich <- c(0.22, 0.29) 
abline( v=HthreshsRich, col="red")
```


```{r, echo=FALSE, fig.height = 5, fig.width = 7}
hetz_poor <- hetzminoo_fancy(my_snpgeno_K1_gl_dapc_e2, 'poor')
HthreshsPoor <- c(0.2225, 0.2775) 
abline( v=HthreshsPoor, col="red")
```

```{r, include=FALSE, echo=FALSE, warning=FALSE}
my_snpgeno_K1_gl_dapc_f <- my_snpgeno_K1_gl_dapc_e2[ (hetz_poor > min(HthreshsPoor)) &
                  (hetz_poor < max(HthreshsPoor)), ]

```

```{r, echo=FALSE, warning=FALSE}
dim(my_snpgeno_K1_gl_dapc_e2) - dim(my_snpgeno_K1_gl_dapc_f) 

```

```{r, include=FALSE, echo=FALSE, warning=FALSE}
my_snpgeno_K1_gl_dapc_f2 <- re_est_ALF(my_snpgeno_K1_gl_dapc_f)
```

```{r, echo=FALSE, fig.height = 5, fig.width = 7}
hetz_poor <- hetzminoo_fancy(my_snpgeno_K1_gl_dapc_f2, "poor")
HthreshsPoor <- c(0.2225, 0.2775) 
abline(v=HthreshsPoor, col="red")
```

```{r, include=FALSE, echo=FALSE, warning=FALSE}
# my_snpgeno_K1_g <- prepare_PLOD_SPA( my_snpgeno_K1_f)
my_snpgeno_K1_gl_dapc_g <- prepare_PLOD_SPA( my_snpgeno_K1_gl_dapc_f2)
names(attributes(my_snpgeno_K1_gl_dapc_g))
ls(my_snpgeno_K1_gl_dapc_g$Kenv)

class(my_snpgeno_K1_gl_dapc_f2)
class(my_snpgeno_K1_gl_dapc_g)
```


```{r, echo=FALSE, fig.height = 7 , fig.width = 10, include=T}
check6and4(my_snpgeno_K1_gl_dapc_f2, thresh_pchisq_6and4=c(0.001, 0.0001))

```

```{r, echo=FALSE, fig.height = 7 , fig.width = 10, include=T}
my_snpgeno_K1_gl_dapc_6_4sf2 <- check6and4(my_snpgeno_K1_gl_dapc_f2, thresh_pchisq_6and4 = c(0.001, 0.00001))
my_snpgeno_K1_gl_dapc_goodLocif2 <- my_snpgeno_K1_gl_dapc_f2[, my_snpgeno_K1_gl_dapc_6_4sf2$pval4 > 0.025]
dim(my_snpgeno_K1_gl_dapc_goodLocif2)
```

```{r}
check6and4(my_snpgeno_K1_gl_dapc_goodLocif2, thresh_pchisq_6and4=c(0.001, 0.0001))

```


```{r}
dim(my_snpgeno_K1_gl_dapc_goodLocif2)
```

```{r}
indiv_filter_CKMR <- my_snpgeno_K1_gl_dapc_goodLocif2$info$Our_sample 
write.csv2(indiv_filter_CKMR,"Pop_structure_filter_kinference.csv",row.names = F)
```

```{r}
# test <- my_snpgeno_K1_gl_dapc_goodLocif2$loci
# write.csv2(test,"Pop_structure_filter_kinference_full.csv",row.names = F)
# 
```


Now we can remove all the individuals that were filter with kinference to our dataset.

```{r}
indiv_keep_CKMR <- read.csv2("Pop_structure_filter_kinference.csv")
colnames(indiv_keep_CKMR) <- "ID_genetique"

# loci_keep_CKMR <- read.csv2("Pop_structure_filter_kinference_full.csv")
# loci_keep_CKMR <- loci_keep_CKMR$AlleleID
# loci_keep_CKMR <- as.data.frame(loci_keep_CKMR)
# colnames(loci_keep_CKMR) <- "AlleleID"

gl_dapc <- gl_dapc[gl_dapc@other$ind.metrics$id %in% indiv_keep_CKMR$ID_genetique, ] 
# dim(gl_dapc)
# gl_dapc_test <- gl_dapc[,gl_dapc@other$loc.metrics$AlleleID %in% loci_keep_CKMR$AlleleID] 
# dim(gl_dapc_test)
# gl_dapc <- gl_dapc_test
```


```{r, echo = F}
dim(gl_dapc)
```

We will delete the samples from the same location with their total number is lower than 10.

```{r}
gl_dapc <- gl_dapc[gl_dapc@other$ind.metrics$pop != "BA",]
gl_dapc <- gl_dapc[gl_dapc@other$ind.metrics$pop != "BD",]
# gl_dapc <- gl_dapc[gl_dapc@other$ind.metrics$pop != "SW",]

dim(gl_dapc)

```


```{r}
gl_dapc_hwe90 <- gl.filter.hwe(gl_dapc,n.pop.threshold = 22) 
gl_dapc_hwe75 <- gl.filter.hwe(gl_dapc,n.pop.threshold = 18) 
gl_dapc_hwe50 <- gl.filter.hwe(gl_dapc,n.pop.threshold = 12) 
gl_dapc_hwe_test <- gl.filter.hwe(gl_dapc,n.pop.threshold = 10) 
gl_dapc_hwe_test1 <- gl.filter.hwe(gl_dapc,n.pop.threshold = 8) 
gl_dapc_hwe_test2 <- gl.filter.hwe(gl_dapc,n.pop.threshold = 6) 
gl_dapc_hwe_test3 <- gl.filter.hwe(gl_dapc,n.pop.threshold = 4) 
gl_dapc_hwe_test4 <- gl.filter.hwe(gl_dapc,n.pop.threshold = 3) 

```


```{r}
threshold_hwe_1 <- data.frame(Nb_pop = c(22,18,12,10,8,6,4,3),Nb_loci=c(2525,2524,2521,2517,2489,2393,2123,1883))

ggplot() + 
  geom_line(mapping = aes(x=threshold_hwe_1$Nb_pop,y=threshold_hwe_1$Nb_loci)) +  
  geom_point(mapping = aes(x=threshold_hwe_1$Nb_pop,y=threshold_hwe_1$Nb_loci))
```


We update the metrics 

```{r, include=FALSE,echo=FALSE, warning=FALSE}
# gl_dapc <- gl.recalc.metrics(gl_dapc)

gl_dapc_hwe_test2 <- gl.filter.hwe(gl_dapc,n.pop.threshold = 8) 
gl_dapc_hwe_test2 <- gl.recalc.metrics(gl_dapc_hwe_test2)

```

```{r, echo=FALSE, fig.height = 5, fig.width = 7}
gl.report.monomorphs(gl_dapc_hwe_test2)

gl_dapc_hwe_test2 <- gl.filter.monomorphs(gl_dapc_hwe_test2)
```

```{r, echo=FALSE, fig.height = 5, fig.width = 7, echo=FALSE, warning=FALSE}
dim(gl_dapc_hwe_test2)

```


```{r}
popNames(gl_dapc_hwe_test2)
```

We transform the genlight object to a genind one which is more frindly to use for next step.
```{r, include=FALSE,echo=FALSE, warning=FALSE}
gl_dapc_hwe_test2_genind <- gl2gi(gl_dapc_hwe_test2)
```

```{r, echo=FALSE, fig.height = 5, fig.width = 7, echo=FALSE, warning=FALSE}
grp_hwe_test2 <- find.clusters(gl_dapc_hwe_test2_genind, max.n.clust=10, n.pca = 600, n.iter = 10000)

```

```{r, include=FALSE,echo=FALSE, warning=FALSE}
names(grp_hwe_test2)
```

```{r,echo=FALSE, warning=FALSE}
head(grp_hwe_test2$Kstat, 20)

```

```{r,echo=FALSE, warning=FALSE}
grp_hwe_test2$stat
```

```{r,echo=FALSE, warning=FALSE}
head(grp_hwe_test2$grp, 10)

```

```{r,echo=FALSE, warning=FALSE}
grp_hwe_test2$size

```

```{r,echo=FALSE, warning=FALSE}
table(pop(gl_dapc_hwe_test2_genind), grp_hwe_test2$grp)

```


```{r}
test_1 <- table(pop(gl_dapc_hwe_test2_genind), grp_hwe_test2$grp)

test_1 <- test_1[c(1,2,6,7,15,8,17,18,19,9,10,11,12,13,14,20,21,22,3,4,5,16),]
```


```{r, include=FALSE,echo=FALSE, warning=FALSE}
location <- c("AA","AC","FA","FB","SA","GL","TB","TD","TF","IA","IB","IC","ID","IE","IF","XA","XB","XC","BB","BC","BE","SW")

```

<!-- # ```{r, echo=FALSE, fig.height = 5, fig.width = 7, echo=FALSE, warning=FALSE} -->
<!-- # table.value(test_1, col.lab=paste("POP", 1:5), row.lab=location) -->
<!-- # -->
<!-- # ``` -->

```{r, include=FALSE,echo=FALSE, warning=FALSE} 
origin_ind_hwe_test2 <- gl_dapc_hwe_test2_genind@other$ind.metrics$pop
indiv_name_hwe_test2 <- gl_dapc_hwe_test2_genind@other$ind.metrics$id
names(origin_ind_hwe_test2) <- indiv_name_hwe_test2
length(origin_ind_hwe_test2)
# This for line allow to add the pop as names of the individuals samples

```

<!-- Allow to test the best number of PCA to retains for DAPC -->
```{r}
x=tab(gl_dapc_hwe_test2_genind,NA.method= "mean") 
crossval= xvalDapc(x,gl_dapc_hwe_test2_genind$pop,result="groupMean",xval.plot=TRUE)
```

```{r}
crossval[2:6]
```

```{r, echo=FALSE, fig.height = 5, fig.width = 7, echo=FALSE, warning=FALSE}
# dapc_hwe_test2 <- dapc(gl_dapc_hwe_test2_genind, origin_ind_hwe_test2, n.pca = 1000, n.da = 3)
dapc_hwe_test2 <- dapc(gl_dapc_hwe_test2_genind, origin_ind_hwe_test2,n.pca = 200, n.da = 5)

```

```{r, echo=FALSE, fig.height = 5, fig.width = 7}
scatter(dapc_hwe_test2, posi.da="bottomright", bg="white", scree.pca=TRUE,posi.pca="bottomleft")

```

```{r}
dapc4= dapc(gl_dapc_hwe_test2_genind,gl_dapc_hwe_test2_genind$pop,n.pca=600,n.da=2)
```

```{r}
percent=dapc4$eig/sum(dapc4$eig)*100 
barplot(percent,ylab="Genetic variance explained by eigenvectors (%)",ylim=c(0,60), names.arg=round(percent,1))
```

<!-- Represent the DAPC using the colors I used on the map -->
```{r}
#Createadata.framecontainingindividualcoordinates 
ind_coords= as.data.frame(dapc4$ind.coord) 
#Renamecolumnsofdataframe 
colnames(ind_coords)= c("Axis1","Axis2") 
#Addacolumncontainingindividuals 
ind_coords$Ind=indNames(gl_dapc_hwe_test2_genind)
#AddacolumnwiththesiteIDs 
ind_coords$Site=gl_dapc_hwe_test2_genind$pop 
#Calculatecentroid(average)positionforeachpopulation 
centroid= aggregate(cbind(Axis1,Axis2)~ Site,data=ind_coords,FUN=mean) 
#Addcentroidcoordinatestoind_coordsdataframe 
ind_coords= left_join(ind_coords,centroid,by="Site",suffix= c("",".cen")) 
#Definecolourpalette c
# ols= brewer.pal(nPop(gl_dapc_genind),"Set2") 
#Customx andylabels 
xlab= paste("Axis1(", format(round(percent[1],1),nsmall=1),"%)",sep="") 
ylab= paste("Axis2(",format(round(percent[2],1),nsmall=1),"%)",sep="")
```

```{r}
ggplot(data=ind_coords, aes(x=Axis1,y=Axis2)) +
  geom_hline(yintercept=0) + 
  geom_vline(xintercept=0) + 
  #spidersegments 
  geom_segment(aes(xend=Axis1.cen, yend=Axis2.cen,colour=Site),show.legend=FALSE) +
  #points 
  geom_point(aes(fill=Site),shape =21,size=3,show.legend=FALSE) +
  #centroids 
  geom_label(data=centroid, aes(label=Site,fill=Site),size=4,show.legend=FALSE) + #colouring 
  # scale_fill_manual(values=cols)+ scale_colour_manual(values=cols)+ #customlabels 
  labs(x=xlab, y=ylab) + ggtitle("DAPC for all samples during all year using Neutral & outliers SNPs") + theme_classic()
```

Now we will made an Fst matrix using Nei equation 

```{r,echo=FALSE, warning=FALSE}
tata <- as.matrix(gl_dapc_hwe_test2_genind$tab)
tata <- as.data.frame(tata)
toto <- gl_dapc_hwe_test2_genind@other$ind.metrics$pop
toto <- as.data.frame(toto)

tato <- cbind(toto,tata)

tato_Fst <- pairwise.neifst(tato,diploid=TRUE) #hierfstat
tato_Fst

```


<!-- Representation of the Fst matrix -->
```{r}
tato_Fst
tato_Fst[lower.tri(tato_Fst)] <- t(tato_Fst)[lower.tri(tato_Fst)]
row.names(tato_Fst) <- colnames(tato_Fst) <- c("AA", "AC","BB", "BC","BE","FA","FB","GL","IA","IB","IC","ID","IE","IF","SA","SW","TB","TD","TF","XA", "XB","XC")
order <- c("AA","AC","FA","FB","SA","GL","TB","TD","TF","IA","IB","IC","ID","IE","IF","XA","XB","XC","BB","BC","BE","SW")
tato_Fst <- reorder_mat(mat = tato_Fst, order = order)
```

```{r, echo=FALSE, fig.height = 5, fig.width = 7}
tato_Fst_matrix <- as.matrix(tato_Fst)

tato_Fst_matrix[lower.tri(tato_Fst_matrix, diag=TRUE)] <- NA
melted <- melt(tato_Fst_matrix, na.rm =TRUE)


ggplot(data = melted, aes(Var2, Var1, fill = value)) + geom_tile(color = "white") + scale_fill_gradient(low = "yellow", high = "blue", name="FST", limits = c(-0.0005,0.005))+ theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 11, hjust = 1),axis.text.y = element_text(size = 12)) + coord_fixed()


# + ggtitle(expression(atop("Pairwise FST, WC (1984)", atop(italic("N = 1694, L = 2024"), "")))) + labs( x = "Sampling Site", y = "Sampling Site")

```

<!-- Calculate the 95% interval for Fst values-->
```{r}
boot.Fst <- boot.ppfst(dat=gl_dapc_hwe_test2_genind,nboot=10000,quant=c(0.025,0.975),diploid=TRUE) 
```


```{r}
boot.Fst_reorder_ll <- as.matrix(boot.Fst$ll)
boot.Fst_reorder_ll[lower.tri(boot.Fst_reorder_ll)] <- t(boot.Fst_reorder_ll)[lower.tri(boot.Fst_reorder_ll)]
row.names(boot.Fst_reorder_ll) <- colnames(boot.Fst_reorder_ll) <- c("AA", "AC","BB","BC","BE","FA","FB","GL","IA","IB","IC","ID","IE","IF", "SA","SW","TB","TD","TF","XA", "XB","XC")
order <- c("AA","AC","FA","FB","SA","GL","TB","TD","TF","IA","IB","IC","ID","IE","IF","XA","XB","XC","BB","BC","BE","SW")
boot.Fst_reorder_ll <- reorder_mat(mat = boot.Fst_reorder_ll, order = order)
```

```{r}
boot.Fst_reorder_ul <- as.matrix(boot.Fst$ul)
boot.Fst_reorder_ul[lower.tri(boot.Fst_reorder_ul)] <- t(boot.Fst_reorder_ul)[lower.tri(boot.Fst_reorder_ul)]
row.names(boot.Fst_reorder_ul) <- colnames(boot.Fst_reorder_ul) <- c("AA", "AC","BB","BC","BE","FA","FB","GL","IA","IB","IC","ID","IE","IF", "SA","SW","TB","TD","TF","XA", "XB","XC")
order <- c("AA","AC","FA","FB","SA","GL","TB","TD","TF","IA","IB","IC","ID","IE","IF","XA","XB","XC","BB","BC","BE","SW")
boot.Fst_reorder_ul <- reorder_mat(mat = boot.Fst_reorder_ul, order = order)
```


```{r}
matrix_95_confident_Fst <- matrix(0, 22, 22)
colnames(matrix_95_confident_Fst) <- colnames(boot.Fst_reorder_ul)
rownames(matrix_95_confident_Fst) <- rownames(boot.Fst_reorder_ul)
matrix_95_confident_Fst[upper.tri(matrix_95_confident_Fst)] <- boot.Fst_reorder_ul[upper.tri(boot.Fst_reorder_ul)]
matrix_95_confident_Fst[lower.tri(matrix_95_confident_Fst, diag=TRUE)] <- NA
matrix_95_confident_Fst[lower.tri(matrix_95_confident_Fst)] <- boot.Fst_reorder_ll[lower.tri(boot.Fst_reorder_ll)]  
  
```

```{r}
tato_95_Fst_matrix <- as.matrix(matrix_95_confident_Fst)

# tato_95_Fst_matrix[lower.tri(tato_95_Fst_matrix, diag=TRUE)] <- NA
test <- melt(tato_95_Fst_matrix)


ggplot(test, aes(x = Var1, y = Var2)) + 
  geom_raster(aes(fill=value)) + 
  scale_fill_gradient(low="yellow", high="blue") +
  labs(x="Sites", y="Sites", title="Matrix 95% interval upper and lower values ") +
  theme_bw() + theme(axis.text.x=element_text(size=9, angle=0, vjust=0.3),
                     axis.text.y=element_text(size=9),
                     plot.title=element_text(size=11))
```


Kinference package :
Bravington MV, Miller DL, Baylis SM (2024). kinference: Pairwise kin-finding from genotypes. R package version 1.0.2






